### ISSUE TITLE
pytorch support

### ISSUE THREAD

**Issue by d4l3k** [NVIDIA/cuda-checkpoint#4]
I just tried this out on PyTorch and it seems to work for the cuda state but I'm hitting issues with criu when saving the parent process. It seems like the issue is with saving the nvidia driver in `criu`.

Are there any plans to expand support for this with `criu` for common ML frameworks?  

```sh
~/D/torch-criu (main)> third-party/cuda-checkpoint/bin/x86_64_Linux/cuda-checkpoint --toggle --pid 125704
~/D/torch-criu (main)> sudo criu dump --shell-job --images-dir demo --tree 125704
[sudo] password for rice: 
Error (criu/files-ext.c:94): Can't dump file 19 of that type [20666] (chr 195:255)
Error (criu/cr-dump.c:1669): Dump files (pid: 125704) failed with -1
Error (criu/cr-dump.c:2093): Dumping FAILED.
```

```sh
~/D/torch-criu (main) [1]> nvidia-smi
Mon Apr 29 17:11:21 2024       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 550.67                 Driver Version: 550.67         CUDA Version: 12.4     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA GeForce RTX 3090        On  |   00000000:06:00.0 Off |                  N/A |
|  0%   32C    P8             16W /  350W |       5MiB /  24576MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
|   1  NVIDIA GeForce RTX 3090        On  |   00000000:07:00.0 Off |                  N/A |
|  0%   34C    P8             21W /  350W |       5MiB /  24576MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
                                                                                         
+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI        PID   Type   Process name                              GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|  No running processes found                                                             |
+-----------------------------------------------------------------------------------------+
```

The file that failed to save seems to be `nvidia`.

```sh
~/D/torch-criu (main)> grep 195 /proc/devices 
195 nvidia
195 nvidia-modeset
195 nvidiactl
```

Test script
```py
import time
import os
import torch

device = torch.device("cuda")

a = torch.tensor(10, device=device)

print(os.getpid())
time.sleep(1000)
```

**Comment by d4l3k** [comment]
Issue seems to specifically be `nvidiactl`

```
python  125704 rice  19u      CHR            195,255        0t0     936 /dev/nvidiactl
python  125704 rice  20u      CHR              195,0        0t0     938 /dev/nvidia0
python  125704 rice  21u      CHR              195,1        0t0     974 /dev/nvidia1
```

**Comment by d4l3k** [comment]
I've also rebuilt PyTorch from source with cuda 12.4.1, cudnn 8.9.7.29-1 and hit the same issue.

**Comment by d4l3k** [comment]
This seems to work on Jax so there's something odd going on with PyTorch

**Comment by d4l3k** [comment]
I hacked around this by modifying `criu` to discard those FDs without errors -- it is able to checkpoint now but I'm not sure how safe it is

**Comment by sgurfinkel** [comment]
This is a known issue and we're working on fixing it!

**Comment by ZingLix** [comment]
> This is a known issue and we're working on fixing it!

This feature is truly helpful! Could you please share if there's a rough timeline or estimated date for this feature to be implemented?

**Comment by ethxnp** [comment]
@sgurfinkel any update on this?

**Comment by thundergolfer** [comment]
@jesus-ramos is the rough timeline on when PyTorch support will land?

**Comment by 913887524gsd** [comment]
> I hacked around this by modifying `criu` to discard those FDs without errors -- it is able to checkpoint now but I'm not sure how safe it is
> 
> diff --git a/criu/files-ext.c b/criu/files-ext.c
> index 95ec8e37c..2a150c546 100644
> --- a/criu/files-ext.c
> +++ b/criu/files-ext.c
> @@ -90,7 +90,9 @@ int dump_unsupp_fd(struct fd_parms *p, int lfd, char *more, char *info, FdinfoEn
> ret = do_dump_gen_file(p, lfd, &ext_dump_ops, e);
> if (ret == 0)
> return 0;
> -   	if (ret == -ENOTSUP)
> +   	if (ret == -ENOTSUP) {
> pr_err("Can't dump file %d of that type [%o] (%s %s)\n", p->fd, p->stat.st_mode, more, info);
> +     	return 0;
> +   	}
> return -1;
> }
> diff --git a/criu/files.c b/criu/files.c
> index 3b653e24b..2ea8ac3ef 100644
> --- a/criu/files.c
> +++ b/criu/files.c
> @@ -847,7 +847,7 @@ int collect_fd(int pid, FdinfoEntry *e, struct rst_info *rst_info, bool fake)
> fdesc = find_file_desc(e);
> if (fdesc == NULL) {
> pr_err("No file for fd %d id %#x\n", e->fd, e->id);
> -           	return -1;
> +           	return 0;
> }
> if (!collect_fd_to(pid, e, rst_info, fdesc, fake, false))

This works fine for fds tied to CUDA devices, but it struggles with PyTorch programs using pinned memory, which is commonly used to speed up data transmission. It's still a bit far from being fully practical...

**Comment by gflarity** [comment]
Is this is still on the roadmap?

**Comment by sgurfinkel** [comment]
> Is this is still on the roadmap?

Yes, it is!

**Comment by lianghao208** [comment]
@sgurfinkel 
Hi, can we know when the next version will be released? And will the code be open source in next release?

**Comment by thundergolfer** [comment]
Hey @sgurfinkel is this fixed on `565.57.01`?

**Comment by sgurfinkel** [comment]
> Hey [@sgurfinkel](https://github.com/sgurfinkel) is this fixed on `565.57.01`?

No, not quite yet!

**Comment by thundergolfer** [comment]
Thanks for the update anyways :) A Google engineer to us indicated it may be fixed on the latest driver. 

@sgurfinkel would NVIDIA be up for doing community-focused video meeting for this project? I'm thinking something similar to what the [AWS Firecracker team did for planning NVIDIA GPU support](https://github.com/firecracker-microvm/firecracker/issues/1179#issuecomment-2374810939).

We (at [modal.com](https://modal.com/)) are very excited about this technology but it's hard to adopt it with little visibility into the system or roadmap :)

**Comment by sgurfinkel** [comment]
> Thanks for the update anyways :) A Google engineer to us indicated it may be fixed on the latest driver.
> 
> [@sgurfinkel](https://github.com/sgurfinkel) would NVIDIA be up for doing community-focused video meeting for this project? I'm thinking something similar to what the [AWS Firecracker team did for planning NVIDIA GPU support](https://github.com/firecracker-microvm/firecracker/issues/1179#issuecomment-2374810939).
> 
> We (at [modal.com](https://modal.com/)) are very excited about this technology but it's hard to adopt it with little visibility into the system or roadmap :)

I'll look into the video meeting, but I do otherwise have an update!
Single-process pytorch support is planned to be released in early 2025!

**Comment by d4l3k** [comment]
@sgurfinkel when you say "single-process" does that mean things like NCCL won't be supported?

**Comment by sgurfinkel** [comment]
> [@sgurfinkel](https://github.com/sgurfinkel) when you say "single-process" does that mean things like NCCL won't be supported?

Yes, that's right.  CUDA IPC support won't be present in the early 2025 release.

**Comment by laochonlam** [comment]
Hi! Do we now have a more concrete release plan for this? Looking forward to it. Thank you!

**Comment by zhuangqh** [comment]
Also looking forward to this ability. Thank you!

**Comment by rst0git** [comment]
@sgurfinkel @jesus-ramos Thank you so much for the update! We were able to confirm checkpoint/restore for PyTorch works as expected with CRIU and driver version 570.86.10.

**Comment by sgurfinkel** [comment]
> [@sgurfinkel](https://github.com/sgurfinkel) [@jesus-ramos](https://github.com/jesus-ramos) Thank you so much for the update! We were able to confirm that checkpoint/restore for PyTorch works as expected with CRIU and driver version 570.86.10.

Yes, NVML support is new in 570!  The readme has been updated to reflect the new features we've added since cuda-checkpoint's initial release!

**Comment by RickyShi46** [comment]
> [@sgurfinkel](https://github.com/sgurfinkel) [@jesus-ramos](https://github.com/jesus-ramos) Thank you so much for the update! We were able to confirm that checkpoint/restore for PyTorch works as expected with CRIU and driver version 570.86.10.

I really appreciate your update. May i ask if there is any tutorial about this function? I also wanna use cuda-checkpoint to checkpoint/restore the pytorch program during model training process.

**Comment by rst0git** [comment]
>May i ask if there is any tutorial about this function? 

https://developer.nvidia.com/blog/checkpointing-cuda-applications-with-criu/

**Comment by amazingguni** [comment]
> > [@sgurfinkel](https://github.com/sgurfinkel) when you say "single-process" does that mean things like NCCL won't be supported?
> 
> Yes, that's right. CUDA IPC support won't be present in the early 2025 release.

In PyTorch, setting num_workers to more than 1 in the `DataLoader` causes an error. It seems to be an issue related to multi-processing, but is this still not supported in the 570.133.07 yet?

![Image](https://github.com/user-attachments/assets/36f04d33-acc7-4942-9df0-7bdb014d0dba)
